pkg/sql/opt/optbuilder/scope_column.go

```diff
@@ -151,3 +156,124 @@ func (*scopeColumn) Eval(_ *tree.EvalContext) (tree.Datum, error) {
 // Variable is part of the tree.VariableExpr interface. This prevents the
 // column from being evaluated during normalization.
 func (*scopeColumn) Variable() {}
+
+// scopeColumnName represents the name of a scopeColumn. The struct tracks two
+// names: refName represents the name that can reference the scopeColumn, while
+// metadataName is the name used when adding the scopeColumn to the metadata.
+//
+// Separating these names allows a column to be referenced by one name while
+// optbuilder builds expressions, but added to the metadata and displayed in opt
+// trees with another. This is useful for:
+//
+//   1. Creating more descriptive metadata names, while having descriptive
+//      refNames that are required for column resolution while building a
+//      expressions. This is particularly useful in mutations where there
+//      are multiple versions of target table columns for fetching,
+//      inserting, and updating that must be referenced by the same name.
+//   2. Creating descriptive metadata names for anonymous columns that
+//      cannot be referenced. This is useful for columns like synthesized
+//      check constraint columns and partial index columns which cannot be
+//      referenced by other expressions. Prior to the creation of
+//      scopeColumnName, the same descriptive name added to the metadata
+//      could be referenced, making optbuilder vulnerable to "ambiguous
+//      column" bugs when a user table had a column with the same name.
+type scopeColumnName struct {
```

Comment:
It might be cleaner to just have these fields in `scopeColumn` with the helper functions below defined on `scopeColumn` too. I originally didn't do that because there are a bunch of places we create `scopeColumn`s with struct literal syntax, e.g. `col := scopeColumn{ ... }`, and I wanted helper functions to create all the different types of names below.

One option is to replace the struct literal syntax with a helper function and setting fields manually afterward:

```diff
- s.cols = append(s.cols, scopeColumn{
-     name:       anonymousScopeColNameWithMetataName("descriptive"),
-     table:      *alias,
-     typ:        tabCol.DatumType(),
-     id:         tabMeta.MetaID.ColumnID(i),
-     visibility: tabCol.Visibility(),
- })
+  col := anonymousScopeColWithMetadataName("descriptive")
+  col.table = *alias
+  col.typ = tabCol.DatumType()
+  col.id = tabMeta.MetaID.ColumnID(i)
+  col.visibility = tabCol.Visibility()
+  s.cols = append(s.cols, col)
```